projects/vdk-plugins/vdk-oracle/src/vdk/plugin/oracle/ingest_to_oracle.py

```diff
@@ -189,40 +189,56 @@ def cast_string_to_type(db_type: str, payload_value: str) -> Any:
             return cast_string_to_type(col_type, value)
         return value
 
-    # TODO: Look into potential optimizations
-    # TODO: https://github.com/vmware/versatile-data-kit/issues/2931
     def _insert_data(self, table_name: str, payload: List[Dict[str, Any]]) -> None:
         if not payload:
             return
 
-        # group dicts by key set
-        batches = {}
-        for p in payload:
-            batch = frozenset(p.keys())
-            if batch not in batches:
-                batches[batch] = []
-            batches[batch].append(p)
-
-        # create queries for groups of dicts with the same key set
-        queries = []
-        batch_data = []
-        for column_names, batch in batches.items():
-            columns = list(column_names)
-            query_columns = [_escape_special_chars(col) for col in columns]
-            insert_sql = f"INSERT INTO {table_name} ({', '.join(query_columns)}) VALUES ({', '.join([':' + str(i + 1) for i in range(len(query_columns))])})"
-            queries.append(insert_sql)
-            temp_data = []
-            for row in batch:
-                temp = [
-                    self._cast_to_correct_type(table_name, col, row[col])
-                    for col in columns
-                ]
-                temp_data.append(temp)
-            batch_data.append(temp_data)
-
-        # batch execute queries for dicts with the same key set
-        for i in range(len(queries)):
-            self.cursor.executemany(queries[i], batch_data[i])
+        batch_size = 1000
+        sorted_rows = []
+        batches = []
+        superset = set()
+
+        # get the set of all column names used in the payload
+        for row in payload:
+            superset.update(row.keys())
+
+        # for each row, fill out the missing columns with None
+        for row in payload:
+            diff = list(superset.difference(row.keys()))
+            for key in diff:
+                row[key] = None
+            # sort each row by column name, so the data passed to executemany is uniform
+            sorted_rows.append(list(dict(sorted(row.items())).values()))
```

Comment:
Not sure about this. We're trading a reduced number of network calls for the overhead of sorting each row by its column names. We should test and possibly revisit this.

Reply:
The way around this is to query all column names from the database, which will always have the same order. We can then fill in the blanks where we have missing keys.

Example https://github.com/vmware/versatile-data-kit/blob/main/projects/vdk-plugins/vdk-postgres/src/vdk/plugin/postgres/ingest_to_postgres.py#L70